File: http/Swift/Sources/Http/Httpclient.swift
// Copyright Core DF — Apache License 2.0

import Foundation

public enum Httpclient {
    private static func parseBody(_ data: Data) -> Any? {
        guard !data.isEmpty else { return nil }
        if let obj = try? JSONSerialization.jsonObject(with: data) { return obj }
        return String(data: data, encoding: .utf8)
    }

    private static func request(_ method: String, url: String, headers: [String: String] = [:], body: Data? = nil) -> [String: Any] {
        var req = URLRequest(url: URL(string: url)!)
        req.httpMethod = method
        req.timeoutInterval = 60
        headers.forEach { req.setValue($0.value, forHTTPHeaderField: $0.key) }
        req.httpBody = body

        let sem = DispatchSemaphore(value: 0)
        var out: [String: Any] = CoreautoResult.transportError()
        URLSession.shared.dataTask(with: req) { data, resp, err in
            defer { sem.signal() }
            if let err {
                out = CoreautoResult.transportError(err.localizedDescription)
                return
            }
            let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
            let parsed = parseBody(data ?? Data())
            if code >= 400 {
                out = ["status_code": code, "error": parsed ?? "inaccessible"]
            } else {
                out = ["status_code": code, "body": parsed as Any]
            }
        }.resume()
        sem.wait()
        return out
    }

    public static func Get(_ url: String, headers: [String: String]? = nil, params: [String: String]? = nil) -> [String: Any] {
        var u = url
        if let params, !params.isEmpty {
            let qs = params.map { "\($0.key)=\($0.value)" }.joined(separator: "&")
            u += (u.contains("?") ? "&" : "?") + qs
        }
        return request("GET", url: u, headers: headers ?? [:])
    }

    public static func Post(_ url: String, jsonBody: Any? = nil, data: String? = nil, headers: [String: String]? = nil) -> [String: Any] {
        var hdrs = headers ?? [:]
        var bodyData: Data?
        if let jsonBody {
            hdrs["Content-Type"] = hdrs["Content-Type"] ?? "application/json"
            bodyData = try? JSONSerialization.data(withJSONObject: jsonBody)
        } else if let data {
            bodyData = data.data(using: .utf8)
        }
        return request("POST", url: url, headers: hdrs, body: bodyData)
    }

    public static func Put(_ url: String, jsonBody: Any? = nil, headers: [String: String]? = nil) -> [String: Any] {
        var hdrs = headers ?? [:]
        var bodyData: Data?
        if let jsonBody {
            hdrs["Content-Type"] = hdrs["Content-Type"] ?? "application/json"
            bodyData = try? JSONSerialization.data(withJSONObject: jsonBody)
        }
        return request("PUT", url: url, headers: hdrs, body: bodyData)
    }

    public static func Delete(_ url: String, headers: [String: String]? = nil) -> [String: Any] {
        request("DELETE", url: url, headers: headers ?? [:])
    }
}

File: http/Go/httpclient/httpclient.go
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Generic HTTP client helpers for Core Auto step scripts (non-Collector REST calls).

package httpclient

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/core-df/coreauto-scripts-pub/http/Go/internal/result"
)

const requestTimeout = 60 * time.Second

var httpClient = &http.Client{Timeout: requestTimeout} // swapped in unit tests

// Get performs an HTTP GET request.
func Get(urlStr string, headers map[string]string, params map[string]string) result.Result {
	if len(params) > 0 {
		u, err := url.Parse(urlStr)
		if err != nil {
			return result.TransportError(err.Error())
		}
		q := u.Query()
		for k, v := range params {
			q.Set(k, v)
		}
		u.RawQuery = q.Encode()
		urlStr = u.String()
	}
	return doRequest(http.MethodGet, urlStr, headers, nil)
}

// Post performs an HTTP POST request with optional JSON or raw body.
func Post(urlStr string, jsonBody any, data []byte, headers map[string]string) result.Result {
	hdrs := cloneHeaders(headers)
	var body []byte
	if jsonBody != nil {
		if _, ok := hdrs["Content-Type"]; !ok {
			hdrs["Content-Type"] = "application/json"
		}
		b, err := json.Marshal(jsonBody)
		if err != nil {
			return result.Result{StatusCode: 500, Error: err.Error()}
		}
		body = b
	} else if data != nil {
		body = data
	}
	return doRequest(http.MethodPost, urlStr, hdrs, body)
}

// Put performs an HTTP PUT request with optional JSON body.
func Put(urlStr string, jsonBody any, headers map[string]string) result.Result {
	hdrs := cloneHeaders(headers)
	var body []byte
	if jsonBody != nil {
		if _, ok := hdrs["Content-Type"]; !ok {
			hdrs["Content-Type"] = "application/json"
		}
		b, err := json.Marshal(jsonBody)
		if err != nil {
			return result.Result{StatusCode: 500, Error: err.Error()}
		}
		body = b
	}
	return doRequest(http.MethodPut, urlStr, hdrs, body)
}

// Delete performs an HTTP DELETE request.
func Delete(urlStr string, headers map[string]string) result.Result {
	return doRequest(http.MethodDelete, urlStr, headers, nil)
}

func doRequest(method, urlStr string, headers map[string]string, body []byte) result.Result {
	var bodyReader io.Reader
	if body != nil {
		bodyReader = bytes.NewReader(body)
	}

	req, err := http.NewRequest(method, urlStr, bodyReader)
	if err != nil {
		return result.TransportError(err.Error())
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return result.TransportError(err.Error())
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return result.TransportError(err.Error())
	}

	parsed := parseBody(respBody)
	if resp.StatusCode >= 400 {
		errVal := any(parsed)
		if parsed == nil {
			errVal = "inaccessible"
		}
		return result.Result{StatusCode: resp.StatusCode, Error: errVal}
	}
	return result.Result{StatusCode: resp.StatusCode, Body: parsed}
}

func parseBody(data []byte) any {
	if len(data) == 0 {
		return nil
	}
	var js any
	if err := json.Unmarshal(data, &js); err == nil {
		return js
	}
	return strings.TrimSpace(string(data))
}

func cloneHeaders(headers map[string]string) map[string]string {
	if headers == nil {
		return map[string]string{}
	}
	out := make(map[string]string, len(headers))
	for k, v := range headers {
		out[k] = v
	}
	return out
}
